ovn-tester/ovn_workload.py

```diff
@@ -553,90 +574,97 @@ def __init__(self, central_node, worker_nodes, cluster_cfg, brex_cfg):
         self.last_selected_worker = 0
         self.n_ns = 0
 
-    def start(self):
-        self.central_node.start(self.cluster_cfg)
+    async def start(self):
+        await self.central_node.start(self.cluster_cfg)
         for w in self.worker_nodes:
```

Comment:
Hi @putnopvut, 
i think here we can start and config workers simultaneously as well 
i tried it with the following change and it's working fine:
```

    async def start(self):
         await self.central_node.start(self.cluster_cfg)
+        start_tasks = []
+        config_tasks = []
+        ctx = ovn_context.active_context
         for w in self.worker_nodes:
-            await w.start(self.cluster_cfg)
-            await w.configure(self.brex_cfg.physical_net)
+            start_tasks.append(ctx.create_task(w.start(self.cluster_cfg)))
+            config_tasks.append(ctx.create_task(w.configure(self.brex_cfg.physical_net)))
+        
+        await asyncio.gather(*start_tasks)
+        await asyncio.gather(*config_tasks2)
```

Reply:
Excellent! I'll fold this change in. Is there a typo in the final line? Should that be `config_tasks` instead of `config_tasks2` ?

Reply:
yes, it is config_tasks :) sorry for the typo

Reply:
Hi Mohammad. I implemented this suggestion, and then I tried it out on a test machine. Even though I can't see any issue with the code, it fails consistently for me.

I added some debug. With the original code, this is what I see happen:
```
2021-09-01 14:06:09,657 | ovn_workload |INFO| Creating localnet on ovn-scale-0
2021-09-01 14:06:09,657 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 14:06:09,744 | ovn_workload |INFO| Adding external host on ovn-scale-0
2021-09-01 14:06:09,744 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 14:06:09,838 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 2
2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| stderr: RTNETLINK answers: File exists

2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 14:06:09,939 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns add ext-ns"
2021-09-01 14:06:10,039 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link set netns ext-ns dev veth0"
2021-09-01 14:06:10,142 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip link set dev veth0 up"
2021-09-01 14:06:10,236 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip addr add 3.0.255.253/16 dev veth0"
2021-09-01 14:06:10,339 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip route add default via 3.0.255.254"
2021-09-01 14:06:10,422 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link set dev veth1 up"
2021-09-01 14:06:10,524 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl add-port br-ex veth1"
```

The `ovn-scale-0` container is created, and a bunch of commands to set up network interfaces is run. Each of those succeeds (with the benign error of adding the same veth0 interface twice).

When I switch to using `asyncio.gather()`, it appears the worker container is not fully started up before we start issuing network commands. Adding some debug, I see results like this:
```
2021-09-01 13:52:32,877 | ovn_workload |INFO| Starting worker ovn-scale-0
2021-09-01 13:52:32,877 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "cd /root/ovn-heater/runtime/ovn-fake-multinode && OVN_MONITOR_ALL=yes OVN_DB_CLUSTER=yes ENABLE_SSL=yes ENABLE_ETCD=no OVN_DP_TYPE=system CREATE_FAKE_VMS=no CHASSIS_COUNT=0 GW_COUNT=0 RELAY_COUNT=3 IP_HOST=192.16.0.0 IP_CIDR=16 IP_START=192.16.0.2 ./ovn_cluster.sh add-chassis ovn-scale-0 tcp:0.0.0.1:6642"
2021-09-01 13:52:32,878 | ovn_workload |INFO| Creating localnet on ovn-scale-0
2021-09-01 13:52:32,878 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 13:52:32,879 | ovn_workload |INFO| Starting worker ovn-scale-1
2021-09-01 13:52:32,879 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "cd /root/ovn-heater/runtime/ovn-fake-multinode && OVN_MONITOR_ALL=yes OVN_DB_CLUSTER=yes ENABLE_SSL=yes ENABLE_ETCD=no OVN_DP_TYPE=system CREATE_FAKE_VMS=no CHASSIS_COUNT=0 GW_COUNT=0 RELAY_COUNT=3 IP_HOST=192.16.0.0 IP_CIDR=16 IP_START=192.16.0.3 ./ovn_cluster.sh add-chassis ovn-scale-1 tcp:0.0.0.1:6642"
2021-09-01 13:52:32,880 | ovn_workload |INFO| Creating localnet on ovn-scale-1
2021-09-01 13:52:32,880 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex exited with status 1
2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,938 | ovn_workload |INFO| Adding external host on ovn-scale-1
2021-09-01 13:52:32,938 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex exited with status 1
2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0

2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,939 | ovn_workload |INFO| Adding external host on ovn-scale-0
2021-09-01 13:52:32,939 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0

2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,984 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,985 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:33,031 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:33,031 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:33,032 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:33,032 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip netns add ext-ns"
2021-09-01 13:52:33,033 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:33,033 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0
...
```
As you can see, it now reports errors about the containers not existing. I'm guessing this is a temporary issue. By the time we move on in the tests and start trying to execute pings on those containers, we no longer get the "no such container" errors. However, since the network interfaces were not set up properly on those containers, it means that the pings fail.

I think we need a more consistent way of knowing the container in question is now up and running. We currently assume that if the SSH command returns, the container is ready. But that appears not to be the case.

As such, I'm going to revert this suggestion for the time being, pending a change for container readiness detection.

Reply:
Hi Mohammad, I figured out the problem. The issue is that `ctx.create_task()` immediately starts the task. So if we create start tasks and configure tasks, they'll both start running ASAP. However, we need to ensure the start tasks complete before we even create the config tasks.

Something like this should work:
```
         await self.central_node.start(self.cluster_cfg)
-         for w in self.worker_nodes:
-            await w.start(self.cluster_cfg)
-            await w.configure(self.brex_cfg.physical_net) 
+        ctx = ovn_context.active_context
+       await asyncio.gather(*[ctx.create_task(w.start(self.cluster_cfg)) for w in self.worker_nodes])
+       await asyncio.gather(*[ctx.create_task(w.configure(self.brex_cfg.physical_net)) for w in self.worker_nodes])
```

Reply:
@putnopvut ohh got it,  great catch :) 
anyway, I think we still can add simultaneously provisioning to other parts of the code :) but for now, maybe you should merge the PRs and will push the other suggestions as a new PRs just to avoid breaking the code again :)